### DISCUSSION TITLE
AVRDUDE warnings

### DISCUSSION THREAD

**Discussion by weiying-chen** [avrdudes/avrdude#1636] [Q&A] [answered]
I get AVRDUDE warnings like these:

>avrdude warning: ATmega329's flash writepage misses a necessary address bit a13 \[/tmp/.avrdude-OHDaQR.conf:5780\]

And these:

>avrdude warning: a4 would normally be expected to be a5 \[/tmp/.avrdude-OHDaQR.conf:11367\]

I think this means there's a discrepancy between my Arduino board and the AVRDUDE config file.

What should I do to fix this? Or maybe this won't cause any major issues since everything seems to running successfully?

    avrdude: 484 bytes of flash verified
    avrdude done.  Thank you. 

Note: I'm using [Arduino Uno R3](https://m.indiamart.com/proddetail/arduino-uno-smd-20433773355.html). And I'm running AVRDUDE using `cargo run` (Rust) on Ubuntu.

Also posted [here](https://www.reddit.com/r/arduino/comments/19dl0cj/avrdude_warnings/).

**Comment by stefanrueger** [comment]
These warnings look like the setup uses a new(er) avrdude with an old configuration file that contains errors. This is not recommended. Always use the avrdude.conf file that comes with the particular release.

If a part is not mentioned by any f these warnings (eg, the ATmega328P) then programming these parts won't be affected.

**Comment by weiying-chen** [comment]
> These warnings look like the setup uses a new(er) avrdude with an old configuration file that contains errors. This is not recommended. Always use the avrdude.conf file that comes with the particular release.
> 
> If a part is not mentioned by any f these warnings (eg, the ATmega328P) then programming these parts won't be affected.

`avrdude -v` outputs this:

```
avrdude: Version 7.1
         Copyright the AVRDUDE authors;
         see https://github.com/avrdudes/avrdude/blob/main/AUTHORS

         System wide configuration file is /etc/avrdude.conf
         User configuration file is /home/wei-ying-chen/.avrduderc
         User configuration file does not exist or is not a regular file, skipping
```
Sorry but where do I get the right `avrdude.conf` and where should I put it?

**Reply by mcuee** [reply]
Check your file `/home/wei-ying-chen/.avrduderc`. You may want to delete it if there is nothing useful.

**Comment by mcuee** [comment]
> Sorry but where do I get the right `avrdude.conf` and where should I put it?

Use the official release here.
https://github.com/avrdudes/avrdude/releases

**Reply by weiying-chen** [reply]
> > Sorry but where do I get the right `avrdude.conf` and where should I put it?
> 
> Use the official release here. https://github.com/avrdudes/avrdude/releases

Thanks for the link.

I installed avrdude with `sudo apt-get install avrdude`. You mean I should uninstall avrdude, then install avrdude again from source using `Source code (tar.gz)`?

**Reply by mcuee** [reply]
Yes that would be a good way.

Or you can use the Arduino binary version here.
https://github.com/arduino/avrdude-packing/releases/tag/7.2-arduino.1

**Reply by weiying-chen** [reply]
I renamed `/usr/local/etc/avrdude.conf` to `/usr/local/etc/avrdude.conf.backup`.

And when I run AVRDUDE with `cargo run`, I get  the same warnings:

```
avrdude warning: ATmega329's flash writepage misses a necessary address bit a13 [/tmp/.avrdude-GzK3P1.conf:5780]
```
But if I check whether AVRDUDE was reading avrdude.conf, it seems to be the case:

```
avrdude main() warning: cannot determine realpath() of config file /usr/local/etc/avrdude.conf: No such file or directory
```

It seems like AVRDUDE is stuck reading those tmp files. I tried deleting one of them ... but then I realized they seem to be created and deleted when the command runs and each time there's new files.

**Reply by weiying-chen** [reply]
I discovered the issue. It's down there, the reply to @stefanrueger. Thanks for helping me!

**Comment by stefanrueger** [comment]
> running AVRDUDE using cargo run (Rust) on Ubuntu

I have never heard of that method. Can you not run avrdude directly from the shell, eg, bash?

**Reply by weiying-chen** [reply, marked as answer]
I found out the problematic config file. It's in the Cargo crate (Rust): [https://github.com/Rahix/avr-hal/blob/main/ravedude/src/avrdude/avrdude.conf](https://github.com/Rahix/avr-hal/blob/main/ravedude/src/avrdude/avrdude.conf)

Thanks for helping me. I have to figure out how to fix that file.

**Reply by stefanrueger** [reply]
You are welcome. Good luck with convincing Cargo crate to use the right .conf file.
